torch/jit.py

```diff
@@ -1,26 +1,496 @@
+#   (that is, we've seen all of the passes, so the trace contains
+#   forwards/backwards/etc), we compile it, and then register this
+#   as the compiled trace.
+# - When we encounter an input configuration whose trace is compiled,
+#   we just directly run the compiled trace.
 #
-# import torch.jit
-# model = model.RNNModel(args.model, ...)
-# model = torch.jit.traced(model)
+# You should never use this class directly; instead, use compile.  However,
+# the intended manual usage of this class looks like this:
+#
+#     class CompiledModel(_CompiledMixin, nn.Module):
+#         def forward(self, x):
+#             ...
+#     CompiledModule.init_compiler()
+#     model = CompiledModule()
+#
+class _CompiledMixin(object):
+    # Global over ALL compilations!  This helps us disambig if two Modules have
+    # the same __name__ but actually are different
+    __next_id = 0
+
+    @classmethod
+    def init_compiler(cls, name=None, enabled=True, time=False, **kwargs):
+        # Ensure we are not shadowing this method on the class we mixed with
+        assert not hasattr(super(_CompiledMixin, cls), "init_compiler")
+        # TODO: Consider saving the backtrace of this constructor, so it's easier
+        # to correlate dump files with invocations in Python
+        #
+        # NB: Use private methods/variables here in order to prevent a class
+        # we mix with from accidentally scrambling us
+        #
+        # NB: Class variables are also accessible via self!
+        kwargs["time"] = time  # also want to pass onto ktrace
+        cls.__ktrace_kwargs = kwargs
+        cls.__enabled = enabled
+        cls.__time = time
+        cls.__model_name = name
+
+        # Monkey patch the constructor and forward functions *inplace*
+        cls.__old_forward = cls.forward
+        cls.forward = cls.__new_forward
+        cls.__old_init = cls.__init__
+        cls.__init__ = cls.__new_init
+
+    def __new_init(self, *args, **kwargs):
+        try:
+            # __old_init is assumed to handle super call
+            self.__old_init(*args, **kwargs)
+        except TypeError as e:
+            # If this fails here, the user probably didn't use this as a class
+            # decorator
+            if "super" in str(e):
+                raise_from(TypeError("torch.jit.compile must be used as a class decorator; "
+                                     "using it on an already defined class is not valid."
+                                     "\n\nOriginal error: {}".format(str(e))), e)
+            else:
+                raise
+        model_name = self.__model_name if self.__model_name else type(self).__name__
+        self.__name = "jit_{}_{}".format(model_name, _CompiledMixin.__next_id)
+        _CompiledMixin.__next_id += 1
+        self.__ktrace_cache = {}
+        self.__next_ktrace_id = 0
+
+    def __process_args(self, args):
+        in_vars, in_struct = _flatten(args, self.state_dict(keep_vars=True).values())
```

Comment:
I assume mutable values used as arguments here will just fail because they are used as keys to a dictionary? I am a bit wary of using python equality of arbitrary data structures in `in_struct` to be used as a key for the traces, but I think we should keep it as is and monitor how it turns out for our examples.

Reply:
Right, if a user passes in a mutable struct, the use of the argument as a key will fail. There are many weird things that I've seen users do, but passing mutable structs to modules is not one of them!